File: Valid Perfect Square - Leetcode 367/Valid Perfect Square - Leetcode 367.ts
// Solution 1: T - O(log n), S - O(1)
const isPerfectSquare = (num: number): boolean => {
  if (num === 0) return false;

  let left = 1,
    right = num;
  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (Math.pow(mid, 2) === num) {
      return true;
    } else if (Math.pow(mid, 2) > num) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }
  return false;
};

File: Valid Perfect Square - Leetcode 367/Valid Perfect Square - Leetcode 367.py
def isPerfectSquare(num:int)->boolean:
    if num === 0:
        return false
    
    left = 0, right = num
    while left <= right:
        mid = (left+right)//2
        if pow(mid,2) === num:
            return true
        elif pow(mid,2) > num:
            right = mid - 1
        else:
            left = mid + 1
    
    return false
